File: src/python/wrapper.py
import asyncio
import importlib.machinery
import importlib.util
import inspect
import json
import os
import sys


class UncachedSourceFileLoader(importlib.machinery.SourceFileLoader):
    """Loads mixed-case Python extensions without sharing a .py bytecode cache."""

    def get_code(self, fullname):
        source_path = self.get_filename(fullname)
        source = self.get_data(source_path)
        return self.source_to_code(source, source_path)


def normalize_module_path(script_path):
    root, extension = os.path.splitext(script_path)
    normalized_extension = extension.lower()
    if extension == normalized_extension or normalized_extension != ".py":
        return script_path

    normalized_path = f"{root}{normalized_extension}"
    try:
        os.lstat(script_path)
        return script_path
    except FileNotFoundError:
        try:
            os.lstat(normalized_path)
            return normalized_path
        except FileNotFoundError:
            return script_path


def call_method(script_path, method_name, *args):
    script_path = normalize_module_path(script_path)
    script_dir = os.path.dirname(os.path.abspath(script_path))
    module_name = os.path.basename(script_path).rsplit(".", 1)[0]
    if script_dir not in sys.path:
        sys.path.insert(0, script_dir)

    print(f"Importing module {module_name} from {script_dir} ...")

    extension = os.path.splitext(script_path)[1]
    if extension.lower() != ".py":
        raise ImportError(f"Cannot load non-Python source file {script_path}")
    loader_class = (
        importlib.machinery.SourceFileLoader
        if extension == ".py"
        else UncachedSourceFileLoader
    )
    loader = loader_class(module_name, script_path)
    spec = importlib.util.spec_from_file_location(
        module_name, script_path, loader=loader
    )
    if spec is None or spec.loader is None:
        raise ImportError(f"Cannot load module from {script_path}")
    script_module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(script_module)

    if "." in method_name:
        class_name, classmethod_name = method_name.split(".")
        method_to_call = getattr(getattr(script_module, class_name), classmethod_name)
    else:
        method_to_call = getattr(script_module, method_name)
    if inspect.iscoroutinefunction(method_to_call):
        return asyncio.run(method_to_call(*args))
    else:
        return method_to_call(*args)


if __name__ == "__main__":
    script_path = sys.argv[1]
    method_name = sys.argv[2]
    json_path = sys.argv[3]
    output_path = sys.argv[4]
    with open(json_path, "r", encoding="utf-8") as fp:
        data = json.load(fp)

    result = call_method(script_path, method_name, *data)
    with open(output_path, "w", encoding="utf-8") as fp:
        # Ensure Unicode is preserved by using ensure_ascii=False
        json.dump({"type": "final_result", "data": result}, fp, ensure_ascii=False)

File: src/ruby/wrapper.rb
# frozen_string_literal: true

require 'json'

##
# Loads the Ruby script at the given path and invokes the named method from that script.
# If `method_name` contains a dot (`ClassName.method`), the class constant is resolved and the
# class method is invoked; otherwise a top-level method is invoked.
# @param [String] script_path - Path to the Ruby script file to load.
# @param [String] method_name - Name of the method to call; use `ClassName.method` to call a class method.
# @param [Array<Object>] args - Arguments to pass to the invoked method.
# @return [Object] The value returned by the invoked method.
def call_method(script_path, method_name, *args)
  script_dir = File.dirname(File.absolute_path(script_path))
  $LOAD_PATH.unshift(script_dir) unless $LOAD_PATH.include?(script_dir)

  puts "Loading script #{script_path} ..."

  # Load the script
  load script_path

  # Call the method
  if method_name.include?('.')
    class_name, method_name_part = method_name.split('.', 2)
    klass = Object.const_get(class_name)
    result = klass.public_send(method_name_part, *args)
  else
    # Call the method using send on self, which has access to top-level methods
    result = send(method_name, *args)
  end

  result
end

if __FILE__ == $PROGRAM_NAME
  script_path = ARGV[0]
  method_name = ARGV[1]
  json_path = ARGV[2]
  output_path = ARGV[3]

  data = JSON.parse(File.read(json_path, encoding: 'utf-8'))

  result = call_method(script_path, method_name, *data)

  File.write(output_path, JSON.generate({ type: 'final_result', data: result }), encoding: 'utf-8')
end
